docs/source/tutorials/memory_optimizations.rst

```diff
@@ -0,0 +1,301 @@
+
+.. code-block:: yaml
+
+  model:
+    apply_lora_to_mlp: True
+    model.lora_attn_modules: ["q_proj", "k_proj", "v_proj"]
+
+Secondly, parameters which control the scale of the impact of LoRA on the model:
+
+* ``lora_rank: int`` affects the scale of the LoRA decomposition, where ``lora_rank << in_dim`` and ``lora_rank << out_dim``
+  \- the dimensions of an arbitrary linear layer in the model. Concretely, ``lora_rank`` reduces the number of gradients stored
+  in a linear fashion from ``in_dim * out_dim`` to ``lora_rank * (in_dim + out_dim)``.
+* ``lora_alpha: float`` affects the magnitude of the LoRA updates. A larger alpha results in larger updates to the base model weights
+  , potentially at the cost of training stability, conversely, smaller alpha can stabilize training at the cost of slower learning.
+  We provide default settings for these parameters which we've tested with all of our models, but we encourage you to adjust them
+  to your specific use case. Typically, one jointly changes ``lora_rank`` and ``lora_alpha`` together.
+* ``lora_dropout`` introduces dropout in the LoRA layers to help regularize training. We default to 0.0 for all of our models.
+
+As above, these parameters are also specified under the ``model`` flag or config entry.
+
+.. note::
+
+  To get a deeper sense of how LoRA parameters affect memory usage during training,
+  see the :ref:`relevant section in our Llama2 LoRA tutorial<lora_tutorial_memory_tradeoff_label>`.
+
+.. _glossary_qlora:
+
+Quantized Low Rank Adaptation (QLoRA)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+*What's going on here?*
+
+You can read our tutorial on :ref:`finetuning Llama2 with QLoRA<qlora_finetune_label>` for a deeper understanding of how it works.
+`QLoRA <https://arxiv.org/abs/2305.14314>`_ is an enhancement on top of `LoRA <https://arxiv.org/abs/2106.09685>`_
+that maintains the frozen model parameters from LoRA in 4-bit quantized precision, thereby reducing memory usage.
+This is enabled through a novel  4-bit NormalFloat (NF4) data type proposed by the authors, which allows for 4-8x less
+parameter memory usage whilst retaining model accuracy.
+
+When considering using QLoRA to reduce memory usage, it's worth noting that QLoRA prevents accuracy degradation during quantization
+by up-casting quantized parameters to the original higher precision datatype during model forward passes - this up-casting may
+incur penalties to training speed. The :ref:`relevant section <qlora_compile_label>` in our QLoRA tutorial demonstrates the usage of ``torch.compile``
+to address this by speeding up training.
+
+*Sounds great! How do I use it?*
+
+You can finetune using QLoRA with any of our LoRA recipes, i.e. recipes with the ``lora_`` prefix, e.g. :ref:`lora_finetune_single_device<lora_finetune_recipe_label>`. These recipes utilize
+QLoRA-enabled model builders, which we support for all our models, and also use the ``qlora_`` prefix, e.g.
```

Comment:
similar to a previous comment, i think that this paragraph can be shorter, and sharing the link to the recipe is probably enough

Reply:
Sorry I think I updated this file - which paragraph were you referring to?